### PR TITLE
[CARROT-903] a possible refinement for JapaneseLanguageComponents

### PR DESCRIPTION
**Pull request by mocobeta**
The fundamental (and, maybe the only one) challenge in https://issues.carrot2.org/browse/CARROT-903 is that word/phrase segmentation is often context-dependent in Japanese texts. The default dictionary bundled in JapaneseTokenizer works well for general ad-hoc search applications (recall is more important in there), but it does not necessarily fit for some higher-level applications such as document clustering. Unfortunately, there are still no de-facto tools or methodologies for that as far as I know;  fine-tuned text split components (and also, domain-specific dictionaries) are often needed to be devised by engineers.

In this PR, I tried to provide an extendable component that hopefully works as an entry point to develop tailored language components for the specific text collection you have. I would not explain the subtle details, but I hope developers who have some experience in tweaking Japanese texts will get the point of the code.

(Appendix)
These are some possible alternatives I can come up with if you are going to get involved:
- use neologd dictionary or Sudachi tokenizer (these are well-known alternatives for the off-the-shelf kuromoji)
- use ShingleFilter to re-construct noun-phrases
- use tailored pre-trained kuromoji system dictionary (i.e. mecab ipadic)  (...if you already have one)

### CODE REVIEW

File: lang/lucene-japanese/src/main/java/org/carrot2/language/japanese/JapaneseLanguageComponents.java [hunk 1]
```diff
@@ -22,12 +38,80 @@
...
+  static {
+    try (Reader reader =
+        new InputStreamReader(
+            JapaneseLanguageComponents.class
+                .getClassLoader()
+                .getResourceAsStream("org/carrot2/language/japanese/japanese.userdict.utf8"),
+            StandardCharsets.UTF_8)) {
+      userDict = UserDictionary.open(reader);
+    } catch (IOException | NullPointerException e) {
+      // TODO
```
Comment (dweiss) [inline lang/lucene-japanese/src/main/java/org/carrot2/language/japanese/JapaneseLanguageComponents.java:52]: I'd throw a runtime exception, even in the patch.

**Review by dweiss — COMMENTED** [review]
Yeah... It looks all right to me. The problem is - I will never be able to tell whether it actually works well. So without your help/ work here it's going to get abandoned. :(  Thank you for leaving the patch though, hopefully somebody is going to find it useful!

**Comment by mocobeta** [comment]
Yes, I wrote the description/explanation not to advocate to merge this into the upstream repo, but to show an example or possibility to developers who want to apply carrot2 on Japanese text collections. This is still a placeholder the same as current `JapaneseLanguageComponents`, and needs to be modified to gain sensible quality.

I think it would be great to tune this on publicly available corpora (e.g., Wikipedia snapshots) so that anyone can see the results and get intuition. But at least, I confirmed carrot2 outputs human-understandable labels for our Japanese corpus with some additional resources/tweaks.

**Comment by dweiss** [comment]
Do you think it'd be good to merge this in in place of what's currently the default? If the default is not working (or working poorly) then your code is a better start for somebody willing to experiment with Japanese...

**Comment by mocobeta** [comment]
I think it can depend on what is expected of the default language component.
I believe this works better as a start point for developers who are ready to get their hands dirty, but the off-the-shelf performance wouldn't be so impressive at all (though it might work slightly better than the current default). So it cannot be used for demo purposes as-is. This is the reason I can't push forward this.

The most critical part of my patch is actually the user dictionary of the kuromoji tokenizer... I cannot share this because the resource I used for testing is proprietary and too domain-specific. If there is a reasonable-sized Japanese corpus (for demo), I could generate the user dictionary resource from that, also make further adjustments. Developing a context-independent tokenizer (for generating good labels) is a difficult task, but we can try to tune it by experiments if a target corpus is given.

**Comment by dweiss** [comment]
Well, anything better than the current default is worth the effort, if you have time to devote to it, it'd be great! Like I said - I rely on you completely to judge the quality of cluster labels. It doesn't have to be perfect - it isn't for English either - but small improvements are always welcome.

**Comment by mocobeta** [comment]
I'm not so confident about pushing this as-is for the reason I described. 
But if - I can find time and create a reasonable-sized (100~1000 docs?) copyright-free text collection to tune the tokenizer with that, is there a way to publish the collection with the tokenizer...? It wouldn't be enough to publish only the adjusted tokenizer itself for trying it out. Can I just include it here (with what format?)

**Comment by dweiss** [comment]
If it's not tangled with incompatible licenses then I think it's fine to include such a collection. If it's a result of some statistical processing (like relationships from Wikipedia terms) then I don't think it's a problem at all - you just give the attribution where it's due.

**Comment by dweiss** [comment]
Hope you're doing well, @mocobeta ! I'm closing this issue, I don't think we'll be pushing this software any much further - there are nicer alternatives these days. It's effectively in maintenance mode only.

**Comment by mocobeta** [comment]
Hi @dweiss, thank you for letting me know that. Understood. Thanks again for your amazing work on this product.
